### PR TITLE
Migrate pallet-im-online to pallet attribute macro.

### PR DESCRIPTION
**Pull request by shaunxw**
Part of #7882.

Converts the `ImOnline` pallet to the new pallet attribute macro introduced in #6877.

Following the upgrade guidelines here: https://crates.parity.io/frame_support/attr.pallet.html#upgrade-guidelines.

## ⚠️ Breaking Change ⚠️ 

From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

> storages now use PalletInfo for module_prefix instead of the one given to decl_storage: Thus any use of this pallet in construct_runtime! should be careful to update name in order not to break storage or to upgrade storage (moreover for instantiable pallet). If pallet is published, make sure to warn about this breaking change.

So users of the `ImOnline` pallet must be careful about the name they used in `construct_runtime!`. Hence the `runtime-migration` label, which might not be needed depending on the configuration of the `ImOnline` pallet.

### CODE REVIEW

File: frame/im-online/src/lib.rs [hunk 1]
```diff
@@ -238,108 +230,152 @@ pub type IdentificationTuple<T> = (
...
+	}
 
-decl_event!(
-	pub enum Event<T> where
-		<T as Config>::AuthorityId,
-		IdentificationTuple = IdentificationTuple<T>,
-	{
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	#[pallet::metadata(T::AuthorityId = "AuthorityId", IdentificationTuple<T> = "IdentificationTuple")]
```
Comment (thiolliere) [inline frame/im-online/src/lib.rs:291]: you need to provide the full type and its associated name
```suggestion
	#[pallet::metadata(T::AuthorityId = "AuthorityId", Vec<IdentificationTuple<T>> = "Vec<IdentificationTuple>")]
```

File: frame/im-online/src/tests.rs [hunk 2]
```diff
@@ -131,7 +131,7 @@ fn heartbeat(
 
 	ImOnline::pre_dispatch(&crate::Call::heartbeat(heartbeat.clone(), signature.clone()))
 		.map_err(|e| match e {
-			TransactionValidityError::Invalid(InvalidTransaction::Custom(INVALID_VALIDATORS_LEN)) =>
+			TransactionValidityError::Invalid(InvalidTransaction::Custom(_)) =>
```
Comment (thiolliere) [inline frame/im-online/src/tests.rs:134]:
```suggestion
			TransactionValidityError::Invalid(InvalidTransaction::Custom(INVALID_VALIDATORS_LEN)) =>
```

File: frame/im-online/src/lib.rs [hunk 3]
```diff
@@ -410,15 +450,69 @@ decl_module! {
 			}
 		}
 	}
-}
 
-type OffchainResult<T, A> = Result<A, OffchainErr<<T as frame_system::Config>::BlockNumber>>;
+	/// Invalid transaction custom error. Returned when validators_len field in heartbeat is incorrect.
+	const INVALID_VALIDATORS_LEN: u8 = 10;
```
Comment (thiolliere) [inline frame/im-online/src/lib.rs:455]:
```suggestion
	pub(crate) const INVALID_VALIDATORS_LEN: u8 = 10;
```

**Review by thiolliere — APPROVED** [review]
the test change is wrong, `INVALID_VALIDATORS_LEN` is matching for a specific value.

and the metadata changed, in general you should compare the metadata before and after the changes to see if storages and types are upgraded correctly.

But apart from that it is good to me

**Review by shawntabrizi — APPROVED** [review]

**Comment by thiolliere** [comment]
bot merge

**Pull request merged**